modules/order/src/MigrateExistingOrderProfiles.php

```diff
@@ -0,0 +1,123 @@
+   * handled differently. The original is migrated to be a billing profile. Then
+   * a copy is created that is migrated to be a shipping profile and the order
+   * shipments that use it are updated to use the new one.
+   *
+   * @param array $order_ids
+   *   An array of order IDs.
+   * @param \Drupal\commerce_order\Entity\OrderTypeInterface $order_type
+   *   The commerce_order_type entity.
+   * @param array $context
+   *   The batch context array.
+   */
+  public static function migrateProfiles(
+    array $order_ids,
+    OrderTypeInterface $order_type,
+    array &$context
+  ) {
+    $message = 'Migrating existing order profiles to use split profiles for
+      billing and shipping...';
+
+    $results = [];
+    foreach ($order_ids as $order_id) {
+      /** @var \Drupal\commerce_order\Entity\OrderInterface $order */
+      $order = \Drupal::entityTypeManager()
+        ->getStorage('commerce_order')
+        ->load($order_id);
+
+      // Grab the billing profile.
+      /** @var \Drupal\profile\Entity\ProfileInterface $billing_profile */
+      $billing_profile = $order->getBillingProfile();
+      if ($billing_profile) {
+        // Change the billing profile from the customer profile type to the
+        // customer_billing profile type.
+        $billing_profile->set('type', 'customer_billing');
+        $billing_profile->save();
+      }
+
+      // Grab the shipping profiles.
+      if ($order->shipments) {
+        foreach ($order->shipments->referencedEntities() as $shipment) {
+          /** @var \Drupal\profile\Entity\ProfileInterface $shipping_profile */
+          $shipping_profile = $shipment->getShippingProfile();
+
+          // We have distinct billing and shipping profile references.
+          if ($shipping_profile->id() != $billing_profile->id()) {
+            // Change the shipping profile from the customer profile type to the
+            // customer_shipping profile type.
+            $shipping_profile->set('type', 'customer_shipping');
+            $shipping_profile->save();
+          }
+          // Else, if the shipment references the same billing profile, create
+          // and save a new duplicate and reference this new shipping profile.
+          else {
+            /** @var \Drupal\profile\Entity\ProfileInterface $new_shipping_profile */
+            $new_shipping_profile = $shipping_profile->createDuplicate();
+
+            // Change the shipping profile from the customer profile type to the
+            // customer_shipping profile type.
+            $new_shipping_profile->set('type', 'customer_shipping');
+            $new_shipping_profile->save();
+
+            /** @var \Drupal\commerce_shipping\Entity\ShipmentInterface $shipment */
+            $shipment->setShippingProfile($new_shipping_profile);
+          }
+        }
+      }
+
+      $results[] = $order->id();
+    }
+
+    $context['message'] = $message;
+    $context['results'] = $results;
+  }
```

Comment:
There is caveat here. A profile can be used elsewhere on other orders as well; that's the point of the addressbook. Patch that enables this is under works. In this code you're only checking if the shipping profile is the same with the billing profile of the current order, but it could be used as the billing profile for another order.

Let's finish the rest for now and we'll improve the logic at the end.

Reply:
I need to apply that patch and get a couple of orders that does that so that I can test it out. I wanted to make sure we are good so far.

Reply:
Could I get a link to that patch?

Reply:
@krystalcode I'm thinking this logic should address this issue:

```diff
diff --git a/modules/order/src/MigrateExistingOrderProfiles.php b/modules/order/src/MigrateExistingOrderProfiles.php
index a92593f7..df05065b 100644
--- a/modules/order/src/MigrateExistingOrderProfiles.php
+++ b/modules/order/src/MigrateExistingOrderProfiles.php
@@ -145,11 +145,60 @@ class MigrateExistingOrderProfiles {
         $new_shipping_profile->set('type', OrderType::PROFILE_SHIPPING);
         $new_shipping_profile->save();

+        // Change the shipment reference to our newly created shipping profile.
         /** @var \Drupal\commerce_shipping\Entity\ShipmentInterface $shipment */
         $shipment->setShippingProfile($new_shipping_profile);
         $shipment->save();
+
+        // Let's also take into account that this profile could have been used
+        // for another order or orders. So, let's grab all other shipments that
+        // referenced this same profile and change the references to our newly
+        // created shipping profile.
+        $this->changeShippingProfileReferences($shipping_profile, $new_shipping_profile);
       }
     }
   }

+  /**
+   * Change the shipping profile reference on orders to the new profile.
+   *
+   * Grab all other orders that reference the original profile and change the
+   * references to our newly created shipping profile.
+   *
+   * @param \Drupal\profile\Entity\ProfileInterface $original_profile
+   *   The original profile that the orders are referencing.
+   * @param \Drupal\profile\Entity\ProfileInterface $new_shipping_profile
+   *   The newly created shipping profile that they should reference.
+   */
+  protected function changeShippingProfileReferences(ProfileInterface $original_profile, ProfileInterface $new_shipping_profile) {
+    $shipments = $this->getReferencedShipments($original_profile->id());
+
+    foreach ($shipments as $shipment) {
+      /** @var \Drupal\commerce_shipping\Entity\ShipmentInterface $shipment */
+      $shipment->setShippingProfile($new_shipping_profile);
+      $shipment->save();
+    }
+  }
+
+  /**
+   * Grab all shipments that reference a profile.
+   *
+   * @param int $profile_id
+   *   The ID of the profile that the shipments are referencing.
+   *
+   * @return array
+   *   Returns an array of shipment entities.
+   */
+  protected function getReferencedShipments($profile_id) {
+    $shipment_storage = \Drupal::entityTypeManager()->getStorage('commerce_shipment');
+    $shipment_ids = $shipment_storage
+      ->getQuery()
+      ->condition('shipping_profile.target_id', $profile_id)
+      ->execute();
+
+    $shipments = $shipment_storage->loadMultiple($shipment_ids);
+
+    return $shipments;
+  }
+
 }
```